File: agent/src/life_update_agent/capture/file_watcher.py
"""Layer 1 - file saves and git commits, via watchdog.

Watches the configured project directories. Directory/file names in
`IGNORED_DIR_NAMES` are skipped entirely (dependency/build noise). A change
to `<repo>/.git/logs/HEAD` is treated as a commit and resolved to its
subject line via `git log`, rather than logged as a plain file write.
"""

from __future__ import annotations

import logging
import subprocess
from datetime import datetime, timezone
from pathlib import Path

from watchdog.events import FileSystemEvent, FileSystemEventHandler
from watchdog.observers import Observer

from life_update_agent import db
from life_update_agent.redaction.scanner import scan

logger = logging.getLogger(__name__)

IGNORED_DIR_NAMES = {
    "node_modules", ".git", "__pycache__", ".venv", "venv", "dist", "build",
    ".next", ".idea", ".vscode", ".pytest_cache", "target", ".turbo",
}


def _is_ignored(path: Path) -> bool:
    return any(part in IGNORED_DIR_NAMES for part in path.parts)


def _is_git_head_log(path: Path) -> bool:
    return path.name == "HEAD" and path.parent.name == "logs" and path.parent.parent.name == ".git"


def _latest_commit_subject(repo_root: Path) -> str | None:
    try:
        result = subprocess.run(
            ["git", "-C", str(repo_root), "log", "-1", "--pretty=%s"],
            capture_output=True, text=True, timeout=5, check=True,
        )
        return result.stdout.strip() or None
    except Exception:
        logger.warning("failed to read latest commit subject for %s", repo_root, exc_info=True)
        return None


def _record(kind: str, file_path: str | None, extra: str | None = None) -> None:
    with db.get_conn() as conn:
        conn.execute(
            "INSERT INTO raw_events (ts, kind, file_path, extra_json) VALUES (?, ?, ?, ?)",
            (datetime.now(timezone.utc).isoformat(), kind, scan(file_path), scan(extra)),
        )


class _Handler(FileSystemEventHandler):
    def on_modified(self, event: FileSystemEvent) -> None:
        self._handle(event)

    def on_created(self, event: FileSystemEvent) -> None:
        self._handle(event)

    def _handle(self, event: FileSystemEvent) -> None:
        if event.is_directory:
            return

        path = Path(str(event.src_path))
        if _is_git_head_log(path):
            repo_root = path.parent.parent.parent
            subject = _latest_commit_subject(repo_root)
            if subject:
                _record("git_commit", str(repo_root), subject)
            return

        if _is_ignored(path) or path.name.startswith("."):
            return

        _record("file", str(path))


def run(watch_dirs: list[str], stop_event) -> None:
    observer = Observer()
    handler = _Handler()
    watched_any = False

    for directory in watch_dirs:
        p = Path(directory).expanduser()
        if not p.is_dir():
            logger.warning("watch dir does not exist, skipping: %s", p)
            continue
        observer.schedule(handler, str(p), recursive=True)
        watched_any = True

    if not watched_any:
        logger.warning("no valid watch directories configured; file watcher idle")
        stop_event.wait()
        return

    observer.start()
    try:
        stop_event.wait()
    finally:
        observer.stop()
        observer.join(timeout=5)

File: app-shell/src-tauri/src/agent/file_watcher.rs
//! Layer 1 - file saves and git commits, via the `notify` crate.
//! A change to `<repo>/.git/logs/HEAD` is treated as a commit and resolved
//! to its subject line via `git log`, rather than logged as a plain write.

use super::{db, redaction::scan, AgentConfig};
use notify::{EventKind, RecursiveMode, Watcher};
use std::path::Path;
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use std::time::Duration;

const IGNORED_DIR_NAMES: &[&str] = &[
    "node_modules", ".git", "__pycache__", ".venv", "venv", "dist", "build",
    ".next", ".idea", ".vscode", ".pytest_cache", "target", ".turbo",
];

fn is_ignored(path: &Path) -> bool {
    path.components().any(|c| {
        let s = c.as_os_str().to_string_lossy();
        IGNORED_DIR_NAMES.contains(&s.as_ref())
    })
}

fn is_git_head_log(path: &Path) -> bool {
    path.file_name().is_some_and(|n| n == "HEAD")
        && path.parent().and_then(|p| p.file_name()).is_some_and(|n| n == "logs")
        && path
            .parent()
            .and_then(|p| p.parent())
            .and_then(|p| p.file_name())
            .is_some_and(|n| n == ".git")
}

fn latest_commit_subject(repo_root: &Path) -> Option<String> {
    let out = std::process::Command::new("git")
        .args(["-C", &repo_root.to_string_lossy(), "log", "-1", "--pretty=%s"])
        .output()
        .ok()?;
    if !out.status.success() {
        return None;
    }
    let subject = String::from_utf8_lossy(&out.stdout).trim().to_string();
    (!subject.is_empty()).then_some(subject)
}

fn handle_path(cfg: &AgentConfig, path: &Path) {
    if is_git_head_log(path) {
        // <repo>/.git/logs/HEAD -> repo root is three levels up.
        if let Some(repo_root) = path.parent().and_then(|p| p.parent()).and_then(|p| p.parent()) {
            if let Some(subject) = latest_commit_subject(repo_root) {
                db::insert_raw(
                    &cfg.db_path,
                    &db::now_iso(),
                    "git_commit",
                    None,
                    None,
                    scan(Some(&repo_root.to_string_lossy())).as_deref(),
                    scan(Some(&subject)).as_deref(),
                );
            }
        }
        return;
    }

    if is_ignored(path) || path.file_name().is_some_and(|n| n.to_string_lossy().starts_with('.')) {
        return;
    }

    db::insert_raw(
        &cfg.db_path,
        &db::now_iso(),
        "file",
        None,
        None,
        scan(Some(&path.to_string_lossy())).as_deref(),
        None,
    );
}

pub fn run(cfg: Arc<AgentConfig>, stop: Arc<AtomicBool>) {
    let (tx, rx) = std::sync::mpsc::channel();
    let mut watcher = match notify::recommended_watcher(move |res| {
        let _ = tx.send(res);
    }) {
        Ok(w) => w,
        Err(e) => {
            log::warn!("file watcher failed to start: {e}");
            return;
        }
    };

    let mut watched_any = false;
    for dir in &cfg.watch_dirs {
        if dir.is_dir() {
            if watcher.watch(dir, RecursiveMode::Recursive).is_ok() {
                watched_any = true;
            }
        } else {
            log::warn!("watch dir does not exist, skipping: {}", dir.display());
        }
    }
    if !watched_any {
        log::info!("no valid watch directories configured; file watcher idle");
        return;
    }

    while !stop.load(Ordering::Relaxed) {
        match rx.recv_timeout(Duration::from_millis(500)) {
            Ok(Ok(event)) => {
                if matches!(event.kind, EventKind::Create(_) | EventKind::Modify(_)) {
                    for path in &event.paths {
                        if path.is_file() {
                            handle_path(&cfg, path);
                        }
                    }
                }
            }
            Ok(Err(_)) | Err(std::sync::mpsc::RecvTimeoutError::Timeout) => {}
            Err(std::sync::mpsc::RecvTimeoutError::Disconnected) => break,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn git_head_log_detection() {
        assert!(is_git_head_log(Path::new("/repo/.git/logs/HEAD")));
        assert!(!is_git_head_log(Path::new("/repo/.git/HEAD")));
        assert!(!is_git_head_log(Path::new("/repo/logs/HEAD")));
    }

    #[test]
    fn ignores_dependency_dirs() {
        assert!(is_ignored(Path::new("/p/node_modules/x/y.js")));
        assert!(is_ignored(Path::new("/p/target/debug/foo")));
        assert!(!is_ignored(Path::new("/p/src/main.rs")));
    }
}
